File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.py
from collections import deque

dr = [1, -1, 0, 0]
dc = [0, 0, 1, -1]

def solution(maps):
    
    n = len(maps)
    m = len(maps[0])
    
    
    def find(sr, sc):
        flag = 0
        cnt = 0
        visited = [[0 for _ in range(m)] for _ in range(n)]
        Q = deque()
        Q.append((sr, sc, cnt, flag, visited))
        
        while Q:
            cr, cc, ccnt, cflag, cvisited = Q.popleft()
            
            if cflag == 0:
                for d in range(4):
                    nr = cr+dr[d]
                    nc = cc+dc[d]
                    if 0<=nr<n and 0<=nc<m and not cvisited[nr][nc]:
                        cvisited[nr][nc] = 1
                        
                        if maps[nr][nc] == "L":
                            Q.append((nr, nc, ccnt+1, 1, cvisited))
                        elif maps[nr][nc] in ("O", "S", "E"):
                            Q.append((nr, nc, ccnt+1, 0, cvisited))
                            
            elif cflag == 1:
                for d in range(4):
                    nr = cr+dr[d]
                    nc = cc+dc[d]
                    if 0<=nr<n and 0<=nc<m and (cvisited[nr][nc] in (0, 1)):
                        cvisited[nr][nc] = 2
                        
                        if maps[nr][nc] == "S":
                            Q.append((nr, nc, ccnt+1, 1, cvisited))
                        elif maps[nr][nc] == "O":
                            Q.append((nr, nc, ccnt+1, 1, cvisited))
                        elif maps[nr][nc] == "E":
                            return ccnt+1
                
    for r in range(n):
        for c in range(m):
            if maps[r][c] == "S":
                return find(r, c) or -1

File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.js
function solution(maps) {
  const dr = [1, -1, 0, 0];
  const dc = [0, 0, 1, -1];
  const n = maps.length;
  const m = maps[0].length;

  function find(sr, sc) {
    const visited = Array.from({ length: n }, () => Array(m).fill(0));
    const Q = [[sr, sc, 0, 0]];        // r, c, cnt, flag
    let head = 0;

    while (head < Q.length) {
      const [cr, cc, cnt, flag] = Q[head++];

      if (flag === 0) {                 // 레버 전
        for (let d = 0; d < 4; d++) {
          const nr = cr + dr[d];
          const nc = cc + dc[d];
          if (0 <= nr && nr < n && 0 <= nc && nc < m && !visited[nr][nc]) {
            visited[nr][nc] = 1;        // 레버 전 방문 표시
            if (maps[nr][nc] === 'L') {
              Q.push([nr, nc, cnt + 1, 1]);
            } else if (['O', 'S', 'E'].includes(maps[nr][nc])) {
              Q.push([nr, nc, cnt + 1, 0]);
            }
          }
        }
      } else {                          // 레버 후
        for (let d = 0; d < 4; d++) {
          const nr = cr + dr[d];
          const nc = cc + dc[d];
          if (
            0 <= nr && nr < n && 0 <= nc && nc < m &&
            (visited[nr][nc] === 0 || visited[nr][nc] === 1)
          ) {
            visited[nr][nc] = 2;        // 레버 후 방문 표시
            if (maps[nr][nc] === 'E') return cnt + 1;
            if (maps[nr][nc] === 'S' || maps[nr][nc] === 'O') {
              Q.push([nr, nc, cnt + 1, 1]);
            }
          }
        }
      }
    }
  }

  for (let r = 0; r < n; r++) {
    for (let c = 0; c < m; c++) {
      if (maps[r][c] === 'S') return find(r, c) ?? -1;
    }
  }
}
